ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,335 @@
+    File fileSizeDir = prepareDirectory(new File(reserveDir, Long.toString(sizeInBytes)));
+    File reserveFile;
+    do {
+      reserveFile = new File(fileSizeDir, generateFilename());
+    } while (!reserveFile.createNewFile());
+    Utils.preAllocateFileIfNeeded(reserveFile, sizeInBytes);
+    return reserveFile;
+  }
+
+  /**
+   * @param requirementsList the list of {@link DiskSpaceRequirements} objects to be accumulated
+   * @return a {@link Map} from segment sizes to the number of reserve segments needed at that size.
+   */
+  private Map<Long, Long> getOverallRequirements(Collection<DiskSpaceRequirements> requirementsList) {
+    Map<Long, Long> overallRequirements = new HashMap<>();
+    Map<Long, Long> swapSegmentsUsed = new HashMap<>();
+
+    for (DiskSpaceRequirements requirements : requirementsList) {
+      long segmentSizeInBytes = requirements.getSegmentSizeInBytes();
+      overallRequirements.put(segmentSizeInBytes,
+          overallRequirements.getOrDefault(segmentSizeInBytes, 0L) + requirements.getSegmentsNeeded());
+      swapSegmentsUsed.put(segmentSizeInBytes,
+          swapSegmentsUsed.getOrDefault(segmentSizeInBytes, 0L) + requirements.getSwapSegmentsInUse());
+    }
+    for (Map.Entry<Long, Long> sizeAndSegmentsNeeded : overallRequirements.entrySet()) {
+      long sizeInBytes = sizeAndSegmentsNeeded.getKey();
+      long origSegmentsNeeded = sizeAndSegmentsNeeded.getValue();
+      // Ensure that swap segments are only added to the requirements if the number of swap segments allocated to
+      // stores is lower than the minimum required swap segments.
+      long swapSegmentsNeeded = Math.max(requiredSwapSegmentsPerSize - swapSegmentsUsed.get(sizeInBytes), 0L);
+      overallRequirements.put(sizeInBytes, origSegmentsNeeded + swapSegmentsNeeded);
+    }
+    return Collections.unmodifiableMap(overallRequirements);
+  }
+
+  /**
+   * Create a directory if it does not yet exist and check that it is accessible.
+   * @param dir The directory to possibly create.
+   * @return the passed-in directory
+   * @throws IOException if the specified directory could not be created/read, or if a non-directory file with the same
+   *                     name already exists.
+   */
+  private static File prepareDirectory(File dir) throws IOException {
+    if (!dir.exists()) {
+      dir.mkdir();
+    }
+    if (!dir.isDirectory()) {
+      throw new IOException("Directory could not be created or is not a directory: " + dir.getAbsolutePath());
+    }
+    return dir;
+  }
+
+  /**
+   * @return a filename for a reserve file. This is reserve_ followed by a random UUID.
+   */
+  private static String generateFilename() {
+    return RESERVE_FILE_PREFIX + UUID.randomUUID();
+  }
+
+  private static class ReserveFileMap {
+    private final Map<Long, Queue<File>> internalMap = new HashMap<>();
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
```

Comment:
would we need the lock if we used a `ConcurrentHashMap` and `LinkedinBlockingQueue`?